rama-tls-acme/src/client.rs

```diff
@@ -0,0 +1,741 @@
+        Ok(nonce)
+    }
+
+    fn get_nonce_from_response(response: &Response) -> Result<String, OpaqueError> {
+        Ok(response
+            .header_str(REPLAY_NONCE_HEADER)
+            .context("get nonce from headers")?
+            .to_owned())
+    }
+
+    /// Create a new account with the given [`CreateAccountOptions`] options
+    pub async fn create_account(
+        &self,
+        ctx: Context<State>,
+        options: CreateAccountOptions,
+    ) -> Result<Account<State>, ClientError> {
+        let key = EcdsaKey::generate().context("generate key for account")?;
+
+        let do_request = async || {
+            let ctx = ctx.clone();
+
+            let response = self
+                .post(ctx, &self.directory.new_account, Some(&options), &key)
+                .await
+                .context("create account request")?;
+
+            let location: String = response
+                .header_str(Location::name())
+                .context("get location header")?
+                .into();
+
+            let account = parse_response::<server::Account>(response).await?;
+            Ok((location, account))
+        };
+
+        let (location, account) = retry_bad_nonce(do_request).await?;
+
+        Ok(Account {
+            client: self,
+            inner: account,
+            credentials: AccountCredentials { key, kid: location },
+        })
+    }
+
+    async fn post(
+        &self,
+        ctx: Context<State>,
+        url: &str,
+        payload: Option<&impl Serialize>,
+        signer: &impl Signer,
+    ) -> Result<Response, OpaqueError> {
+        let mut builder = JWSBuilder::new();
+        if let Some(payload) = payload {
+            let payload = serde_json::to_vec(payload).context("serialize payload")?;
+            builder.set_payload(payload);
+        }
+
+        let nonce = if let Some(nonce) = self.nonce.lock().take() {
+            nonce
+        } else {
+            self.fetch_nonce(ctx.clone()).await?
+        };
+
+        builder.try_set_protected_headers(ProtectedHeaderAcme {
+            nonce: Cow::Owned(nonce),
+            url: Cow::Borrowed(url),
+        })?;
+
+        let jws = builder.build_flattened(signer)?;
+
+        let request = self
+            .https_client
+            .post(url)
+            .typed_header(UserAgent::from_static("rama-tls-acme"))
+            .typed_header(ContentType::jose_json())
+            .json(&jws);
+
+        let response = request.send(ctx).await?;
+
+        *self.nonce.lock() = Some(Self::get_nonce_from_response(&response)?);
+        Ok(response)
+    }
+}
+
+#[derive(Clone, Debug)]
+/// Enum of popular acme providers and their directory url
+pub enum AcmeProvider {
+    LetsEncryptProduction,
+    LetsEncryptStaging,
+    ZeroSslProduction,
+    GoogleTrustServicesProduction,
+    GoogleTrustServicesStaging,
+}
+
+impl AcmeProvider {
+    #[must_use]
+    pub fn as_str(&self) -> &str {
```

Comment:
as_str is very unclear of the meaning of the string here. Would rename it as `as_endpoint` or `as_uri` or something like that